File: CashuWallet/Core/Protocols/CurrencyProtocol.swift
import Foundation

// MARK: - Currency Protocol

/// Protocol defining the interface for different currencies.
/// Implementations include SAT (satoshis), USD, EUR, and potentially other fiat currencies.
///
/// This abstraction allows the wallet to:
/// - Display amounts in different currency denominations
/// - Support mints with different unit types
/// - Convert between currencies when needed
protocol Currency {
    /// ISO-style currency code (e.g., "SAT", "USD", "EUR")
    var code: String { get }
    
    /// Currency symbol for display (e.g., "₿", "$", "€")
    var symbol: String { get }
    
    /// Number of decimal places for this currency
    /// SAT = 0, USD/EUR = 2
    var decimals: Int { get }
    
    /// Human-readable name (e.g., "Satoshis", "US Dollar", "Euro")
    var displayName: String { get }
    
    /// Whether symbol appears before or after the amount
    var symbolPosition: CurrencySymbolPosition { get }
}

/// Position of currency symbol relative to amount
enum CurrencySymbolPosition {
    case before  // $100
    case after   // 100 SAT
}

// MARK: - Built-in Currencies

/// Satoshis - the base unit for Bitcoin/Lightning
struct SatoshiCurrency: Currency {
    let code = "SAT"
    let symbol = "₿"
    let decimals = 0
    let displayName = "Satoshis"
    let symbolPosition = CurrencySymbolPosition.before
}

/// US Dollar
struct USDCurrency: Currency {
    let code = "USD"
    let symbol = "$"
    let decimals = 2
    let displayName = "US Dollar"
    let symbolPosition = CurrencySymbolPosition.before
}

/// Euro
struct EURCurrency: Currency {
    let code = "EUR"
    let symbol = "€"
    let decimals = 2
    let displayName = "Euro"
    let symbolPosition = CurrencySymbolPosition.before
}

/// Fallback for an arbitrary mint unit the registry doesn't know about (any
/// custom unit string a mint might advertise). Treated as an integer base unit
/// (no decimals) and displayed with the raw code as a trailing label.
struct GenericCurrency: Currency {
    let code: String
    let symbol = ""
    let decimals = 0
    let displayName: String
    let symbolPosition = CurrencySymbolPosition.after

    init(unit: String) {
        let normalized = unit.uppercased()
        self.code = normalized
        self.displayName = normalized
    }
}

// MARK: - Currency Amount

/// A value with an associated currency
struct CurrencyAmount: Equatable {
    /// The raw value in the smallest unit (sats for BTC, cents for fiat)
    let value: UInt64
    
    /// The currency this amount is denominated in
    let currency: any Currency
    
    /// Create amount from smallest unit value
    init(value: UInt64, currency: any Currency) {
        self.value = value
        self.currency = currency
    }
    
    /// Create satoshi amount (convenience)
    static func sats(_ value: UInt64) -> CurrencyAmount {
        CurrencyAmount(value: value, currency: SatoshiCurrency())
    }
    
    /// Create USD amount from cents
    static func usdCents(_ cents: UInt64) -> CurrencyAmount {
        CurrencyAmount(value: cents, currency: USDCurrency())
    }
    
    /// Create EUR amount from cents
    static func eurCents(_ cents: UInt64) -> CurrencyAmount {
        CurrencyAmount(value: cents, currency: EURCurrency())
    }
    
    /// The display value (e.g., 1000 cents = 10.00)
    var displayValue: Double {
        if currency.decimals == 0 {
            return Double(value)
        }
        return Double(value) / pow(10.0, Double(currency.decimals))
    }
    
    /// Formatted string for display
    func formatted(showSymbol: Bool = true) -> String {
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.numberStyle = .decimal
        formatter.minimumFractionDigits = currency.decimals
        formatter.maximumFractionDigits = currency.decimals
        formatter.usesGroupingSeparator = true
        formatter.groupingSeparator = ","
        formatter.decimalSeparator = "."
        
        let formattedValue = formatter.string(from: NSNumber(value: displayValue)) ?? "\(displayValue)"
        
        guard showSymbol else { return formattedValue }
        
        switch currency.symbolPosition {
        case .before:
            return "\(currency.symbol)\(formattedValue)"
        case .after:
            return "\(formattedValue) \(currency.code)"
        }
    }
    
    // MARK: - Equatable
    
    static func == (lhs: CurrencyAmount, rhs: CurrencyAmount) -> Bool {
        lhs.value == rhs.value && lhs.currency.code == rhs.currency.code
    }
}

// MARK: - Currency Registry

/// Registry for looking up currencies by code
enum CurrencyRegistry {
    /// All supported currencies
    static let supportedCurrencies: [any Currency] = [
        SatoshiCurrency(),
        USDCurrency(),
        EURCurrency()
    ]
    
    /// Get currency by code
    static func currency(forCode code: String) -> (any Currency)? {
        supportedCurrencies.first { $0.code.uppercased() == code.uppercased() }
    }
    
    /// Map a mint unit string to a currency for entry precision + display.
    /// Known units resolve to their built-in currency; any other (custom) unit
    /// falls back to a `GenericCurrency` so the result is never nil and
    /// arbitrary mint units are supported.
    static func currency(forMintUnit unit: String) -> any Currency {
        switch unit.lowercased() {
        case "sat", "sats", "satoshi", "satoshis":
            return SatoshiCurrency()
        case "usd", "dollar", "dollars":
            return USDCurrency()
        case "eur", "euro", "euros":
            return EURCurrency()
        default:
            return GenericCurrency(unit: unit)
        }
    }
}

File: android/app/src/main/java/com/cashu/me/Core/Protocols/CurrencyProtocol.kt
package com.cashu.me.Core.Protocols

import java.text.NumberFormat
import java.util.Locale
import kotlin.math.pow

interface CurrencyDisplay {
    fun formatSats(amount: Long, includeUnit: Boolean = true): String
    fun formatFiat(amountSats: Long, btcPrice: Double?, currencyCode: String): String?
}

enum class CurrencySymbolPosition {
    Before,
    After,
}

data class WalletCurrency(
    val code: String,
    val symbol: String,
    val decimals: Int,
    val displayName: String,
    val symbolPosition: CurrencySymbolPosition,
)

object WalletCurrencies {
    val Satoshi = WalletCurrency(
        code = "SAT",
        symbol = "₿",
        decimals = 0,
        displayName = "Satoshis",
        symbolPosition = CurrencySymbolPosition.Before,
    )
    val Usd = WalletCurrency(
        code = "USD",
        symbol = "$",
        decimals = 2,
        displayName = "US Dollar",
        symbolPosition = CurrencySymbolPosition.Before,
    )
    val Eur = WalletCurrency(
        code = "EUR",
        symbol = "€",
        decimals = 2,
        displayName = "Euro",
        symbolPosition = CurrencySymbolPosition.Before,
    )
}

data class CurrencyAmount(
    val value: Long,
    val currency: WalletCurrency,
) {
    init {
        require(value >= 0) { "Currency amount must be non-negative." }
        require(currency.decimals >= 0) { "Currency decimals must be non-negative." }
    }

    val displayValue: Double
        get() = if (currency.decimals == 0) {
            value.toDouble()
        } else {
            value.toDouble() / 10.0.pow(currency.decimals.toDouble())
        }

    fun formatted(showSymbol: Boolean = true, locale: Locale = Locale.US): String {
        val formatter = NumberFormat.getNumberInstance(locale).apply {
            minimumFractionDigits = this@CurrencyAmount.currency.decimals
            maximumFractionDigits = this@CurrencyAmount.currency.decimals
            isGroupingUsed = true
        }
        val formattedValue = formatter.format(displayValue)
        if (!showSymbol) return formattedValue
        return when (currency.symbolPosition) {
            CurrencySymbolPosition.Before -> "${currency.symbol}$formattedValue"
            CurrencySymbolPosition.After -> "$formattedValue ${currency.code}"
        }
    }

    companion object {
        fun sats(value: Long): CurrencyAmount = CurrencyAmount(value, WalletCurrencies.Satoshi)
        fun usdCents(cents: Long): CurrencyAmount = CurrencyAmount(cents, WalletCurrencies.Usd)
        fun eurCents(cents: Long): CurrencyAmount = CurrencyAmount(cents, WalletCurrencies.Eur)
    }
}

object CurrencyRegistry {
    val supportedCurrencies: List<WalletCurrency> = listOf(
        WalletCurrencies.Satoshi,
        WalletCurrencies.Usd,
        WalletCurrencies.Eur,
    )

    fun currencyForCode(code: String): WalletCurrency? {
        val normalized = code.trim()
        return supportedCurrencies.firstOrNull { it.code.equals(normalized, ignoreCase = true) }
    }

    /**
     * Never null: unknown mint units resolve to a generic code-after currency
     * (0 decimals, uppercase code as the unit label), mirroring iOS
     * GenericCurrency so custom-unit mints always format.
     */
    fun currencyForMintUnit(unit: String): WalletCurrency {
        return when (unit.trim().lowercase()) {
            "sat", "sats", "satoshi", "satoshis" -> WalletCurrencies.Satoshi
            "usd", "dollar", "dollars" -> WalletCurrencies.Usd
            "eur", "euro", "euros" -> WalletCurrencies.Eur
            else -> genericCurrency(unit)
        }
    }

    private fun genericCurrency(unit: String): WalletCurrency {
        val code = unit.trim().uppercase().ifEmpty { "SAT" }
        return WalletCurrency(
            code = code,
            symbol = "",
            decimals = 0,
            displayName = code,
            symbolPosition = CurrencySymbolPosition.After,
        )
    }
}
